azure-iot-device/azure/iot/device/iothub/sync_clients.py

```diff
@@ -183,6 +183,62 @@ def callback():
         self._transport.enable_feature(feature_name, callback=callback)
         enable_complete.wait()
 
+    def get_twin(self):
+        """
+        Gets the device or module twin from the Azure IoT Hub or Azure IoT Edge Hub service.
+
+        This is a synchronous call, meaning that this function will not return until the twin
+        has been retrieved from the service.
+
+        :returns: Twin object which was retrieved from the hub
+        """
+        # OPEN QUESTON: does this have a timeout parameter and does it raise on timeout?
```

Comment:
>OPEN QUESTON [](start = 10, length = 12)

I think it should have a time out.

Reply:
This definitely needs to follow the same pattern as the receive APIs - it needs both the `block` and `timeout` optional params.

---
In reply to: [287531765](https://github.com/Azure/azure-iot-sdk-python-preview/pull/96#discussion_r287531765) [](ancestors = 287531765)

Reply:
For block,  that parameter is meaningless because this is a case where we're sending a request to the hub and waiting for a response.  Setting block=False would effectively say that the library needs to send a request but not wait for the response.

For timeout, I agree that we need a mechanism for this, but timeout needs to apply to almost all operations and the timeout for get_twin is different than the timeout for receive_method_request.  In receive_method-request, a timeout means "nothing is wrong, but we didn't receive anything".  In get_twin, a timeout means "I tried to get the twin, but it didn't come back.  Something is wrong."

Timeout also implies that we need a cancellation mechanism that we don't have.  If a get_twin operation times out, there should be a way to cancel that operation, and that doesn't exist yet.


---
In reply to: [288237256](https://github.com/Azure/azure-iot-sdk-python-preview/pull/96#discussion_r288237256) [](ancestors = 288237256,287531765)

Reply:
Yeah I was working on editing the comment when you replied. Ignore the previous comment, per our discussion.

---
In reply to: [288271386](https://github.com/Azure/azure-iot-sdk-python-preview/pull/96#discussion_r288271386) [](ancestors = 288271386,288237256,287531765)